domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,209 @@
+            switchIconificationCount
+          )
+          is CommandMessage.TakeSnapshot -> {
+            if (message.switchId == switchIconificationCount) {
+              val newSnapshot =
+                performanceMetricsAssessor.computeSnapshotAtCurrentTime(currentIconification)
+              sendLogSnapshotDiffCommand(
+                performanceMetricsAssessor.getRelativeCpuUsage(previousSnapshot, newSnapshot)
+              )
+              previousSnapshot = newSnapshot
+              sendScheduleTakeSnapshotCommand(currentIconification, switchIconificationCount)
+            }
+          }
+          is CommandMessage.LogSnapshotDiff -> {
+            performanceMetricsLogger.logCpuUsage(
+              currentIconification.toScreenName(),
+              message.relativeCpuUsage
+            )
+          }
+        }
+      }
+    }.also {
+      CoroutineScope(backgroundCoroutineDispatcher).launch {
+        sendTakeSnapshotCommand(switchIconificationCount)
+      }
+    }
+  }
+
+  private fun sendSwitchIconificationCommand(newIconification: AppIconification) {
+    commandQueue.offer(CommandMessage.SwitchIconification(newIconification)).apply {
+      if (!this) {
+        val exception = IllegalStateException("Iconification switching failed")
+        consoleLogger.e(
+          "CpuPerformanceSnapshotter",
+          "Failure while switching AppIconification.",
+          exception
+        )
+        exceptionLogger.logException(exception)
+      }
+    }
+  }
+
+  private suspend fun sendScheduleTakeSnapshotCommand(
+    currentIconification: AppIconification,
+    switchId: Int
+  ) { commandQueue.send(CommandMessage.ScheduleTakeSnapshot(currentIconification, switchId)) }
+
+  private suspend fun sendTakeSnapshotCommand(switchId: Int) {
+    commandQueue.send(CommandMessage.TakeSnapshot(switchId))
+  }
+
+  private suspend fun sendLogSnapshotDiffCommand(relativeCpuUsage: Double) {
+    commandQueue.send(CommandMessage.LogSnapshotDiff(relativeCpuUsage))
+  }
+
+  /**
+   * Schedules a delay on the basis of [currentIconification] and then sends the [CommandMessage]
+   * for taking a [Snapshot] that'll be used for logging the relative cpu-usage of the application.
+   */
+  private fun scheduleTakeSnapshot(currentIconification: AppIconification, switchId: Int) {
+    CoroutineScope(backgroundCoroutineDispatcher).launch {
+      delay(currentIconification.getDelay())
+      sendTakeSnapshotCommand(switchId)
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          "PerformanceMetricsController",
+          "Failed to remove metric log.",
+          it
+        )
+      }
+    }
+  }
+
+  /**
+   * Represents a message that can be sent to [commandQueue] to process cpu-usage logging.
+   *
+   * Messages are expected to be resolved serially (though their scheduling can occur across
+   * multiple threads, so order cannot be guaranteed until they're enqueued).
+   */
+  sealed class CommandMessage {
```

Comment:
Can this be private? Prefer isolating its availability since it probably shouldn't be needed outside this class (it's an implementation detail of the class).